runtime/include/QuantumDevice.hpp

```diff
@@ -55,310 +57,558 @@ struct QuantumDevice {
     QuantumDevice &operator=(QuantumDevice &&) = delete;
 
-    /**
-     * @brief Allocate a qubit.
-     *
-     * @return `QubitIdType`
-     */
-    virtual auto AllocateQubit() -> QubitIdType = 0;
+    // ----------------------------------------
+    //  QUBIT MANAGEMENT
+    // ----------------------------------------
 
     /**
-     * @brief Allocate a vector of qubits.
+     * @brief Allocate an array of qubits.
+     *
+     * The operation should perform the necessary steps to make the given number of qubits
+     * available in a clean |0> state. At minimum, a device must be able to execute this
+     * call once before any quantum operations have been run. Handling multiple allocation calls
+     * is optional and only useful if the device intends to support dynamic qubit allocations
+     * (that is allocations during quantum program execution).
+     *
+     * While devices may choose not to distinguish between logical and device IDs, having a logical
+     * qubit labeling system can help catch program errors such as addressing a previously released
+     * qubit. An example implementation can be found in `QubitManager.hpp`. Devices are also free
+     * to disable a resource without physically freeing it, allowing for faster dynamic allocations.
+     * The returned IDs must be accepted in addressing operations (like gates) until the quantum
+     * resource is explicitly freed or the device is destroyed.
      *
      * @param num_qubits The number of qubits to allocate.
      *
-     * @return `std::vector<QubitIdType>`
+     * @return `std::vector<QubitIdType>` Array of logical qubit IDs.
      */
     virtual auto AllocateQubits(size_t num_qubits) -> std::vector<QubitIdType> = 0;
 
     /**
-     * @brief Release a qubit.
+     * @brief Release all qubits and reset device state.
+     *
+     * This operation can be considered a device-wide quantum resource release, regardless of how
+     * many allocation calls have been run (although on devices with static allocation there would
+     * only be one matching `AllocateQubits` - `ReleaseAllQubits` pair).
+     * After executing this call, a device should be in a state ready to allocate new qubits and
+     * execute another quantum program. Configurable execution parameters, whether explicit in
+     * the interface (like shots) or internal via the constructor kwargs, should be maintained
+     * in the same state as when entering this method.
+     */
+    virtual void ReleaseAllQubits() = 0;
+
+    /**
+     * @brief Get the number of currently allocated qubits.
      *
-     * @param qubit The id of the qubit
+     * @return `size_t`
      */
-    virtual void ReleaseQubit(QubitIdType qubit) = 0;
+    virtual auto GetNumQubits() const -> size_t = 0;
 
     /**
-     * @brief Release all qubits.
+     * @brief (Optional) Allocate a qubit.
+     *
+     * Allocate a new qubit in the |0> state. This method is only needed for devices with dynamic
```

Comment:
Could you also mention that the behaviour of adding a new qubit (little-endian or big-endian order) to the state is left up to the device?

Reply:
That sort of language seems highly implementation specific to me, unless the concept of endianness comes up anywhere else in the device?